otp/imx6ull-otp/imx6ull-otp.c

```diff
@@ -549,6 +557,48 @@ static int write_sn(const char *sn, int raw)
 }
 
 
+static int read_hwrev(int silent)
+{
+	uint32_t plc1 = *(otp.base + ocotp_gp2);
+
+	// NOTE: keeping (hw_rev - 1) in FUSE bits
+	uint8_t hw_rev = (plc1 >> 24 & 0xff) + 1;
+	if (silent == 0) {
+		printf("%u\n", hw_rev);
+	}
+
+	return hw_rev;
+}
+
+
+static int write_hwrev(uint8_t hw_rev)
+{
+	int ret = 0;
+
+	if ((read_hwrev(1) != 1) && !(common.force == 0)) {
+		printf("HW rev already written\n");
+		return -1;
+	}
+
+	// NOTE: no need to read current plc1 word, HW does it automatically
```

Comment:
Hmm, I guess you mean that writing zeroes does not burn OTP (0 is unburned state)?

Reply:
reference manual explicitly states that it's being read to mask the current value:
and address can be written in the same operation.

> Write the data to the HW_OCOTP_DATA register. This will automatically set
> HW_OCOTP_CTRL[BUSY] and clear HW_OCOTP_CTRL[WR_UNLOCK]. To
> protect programming same OTP bit twice, before program OCOTP will
> automatically read fuse value in OTP and use read value to mask program data. The
> controller will use masked program data to program a 32-bit word in the OTP per the
> address in HW_OCOTP_CTRL[ADDR]. Bit fields with 1's will result in that OTP
> bit being programmed. Bit fields with 0's will be ignored.